tty.go

```diff
@@ -99,18 +100,32 @@ func (t *tty) recvtty(process *libcontainer.Process, socket *os.File) (Err error
 	t.wg.Add(1)
 	go t.copyIO(os.Stdout, epollConsole)
 
-	// set raw mode to stdin and also handle interrupt
-	stdin, err := console.ConsoleFromFile(os.Stdin)
-	if err != nil {
-		return err
+	// Set raw mode for the controlling terminal. Usually all three
+	// (stdin, stdout, and stderr) streams are open to the terminal,
+	// but they might be redirected, so try them all.
+	var hostConsole console.Console
+	for _, s := range []*os.File{os.Stderr, os.Stdout, os.Stdin} {
+		c, err := console.ConsoleFromFile(s)
+		switch err {
+		case console.ErrNotAConsole:
+			continue
+		case nil:
+			hostConsole = c
+		default:
+			return errors.Wrap(err, "unable to get console")
+		}
 	}
-	if err := stdin.SetRaw(); err != nil {
+	if hostConsole == nil {
+		return errors.New("unable to find console (are all streams redirected?)")
```

Comment:
I think a nicer solution might be to just not do interrupt or resize handling if there isn't a console, but still allow the container to run. I don't really have an opinion on whether we should check `stdout` or `stderr` ... it feels a *little* wrong because `Ctrl-C` feels like an "input" thing but :shrug:.

Reply:
We need a terminal to set the raw mode etc -- and any terminal we have will do. Tools like `clear` and `tput` do the same thing -- they try stderr, stdout, stdin, and if all three fails, open("/dev/tty") should do the same thing.

I didn't do open("/dev/tty") as I am not sure where to close it, but I think I'll implement that later, to make cases like`runc run $ID top </dev/null >/dev/null 2>&1`work.

Reply:
> not do interrupt or resize handling if there isn't a console, but still allow the container to run

I disagree. If we say in config we want a console, and there's no [host] console we should bail out. This is what my current code does (see https://github.com/opencontainers/runc/pull/2535#issuecomment-667427446)

Reply:
@cyphar PTAL ^^^